xla/service/gpu/gpu_reduce_scatter_all_gather_combiner.cc

```diff
@@ -0,0 +1,110 @@
+/* Copyright 2023 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/compiler/xla/service/gpu/gpu_reduce_scatter_all_gather_combiner.h"
+
+#include "tensorflow/compiler/xla/hlo/ir/hlo_casting_utils.h"
+#include "tensorflow/compiler/xla/hlo/ir/hlo_instruction.h"
+#include "tensorflow/compiler/xla/hlo/ir/hlo_instructions.h"
+#include "tensorflow/compiler/xla/hlo/ir/hlo_module.h"
+#include "tensorflow/compiler/xla/hlo/ir/hlo_opcode.h"
+#include "tensorflow/compiler/xla/service/collective_ops_utils.h"
+#include "tensorflow/compiler/xla/service/hlo_query.h"
+
+namespace xla {
+namespace gpu {
+
+StatusOr<bool> ReduceScatterAllGatherCombiner::Run(
+    HloModule* module,
+    const absl::flat_hash_set<absl::string_view>& execution_threads) {
+  const HloModuleConfig& config = module->config();
+  int64_t next_channel_id = hlo_query::NextChannelId(*module);
+
+  bool changed = false;
+  const int64_t min_rank = 1;
+  for (HloComputation* computation :
+       module->MakeNonfusionComputations(execution_threads)) {
+    for (HloInstruction* instruction :
+         computation->MakeInstructionPostOrder()) {
+      if (instruction->opcode() != HloOpcode::kReduceScatter ||
+          !instruction->shape().IsArray()) {
+        continue;
+      }
+      auto* rs = Cast<HloReduceScatterInstruction>(instruction);
+
+      if (!rs->shape().IsArray() || rs->constrain_layout() ||
+          (rs->IsCrossModuleAllReduce() &&
+           !rs->GetModule()->config().use_spmd_partitioning())) {
+        VLOG(2) << "Unsupported reduce-scatter: " << rs->ToString();
+        continue;
+      }
+      if (rs->shape().rank() - absl::c_count(rs->shape().dimensions(), 1) <
+          min_rank) {
+        VLOG(2) << " Should be at least rank-" << min_rank
+                << " excluding trivial dimensions " << rs->ToString();
+        continue;
+      }
+      if (rs->replica_groups().size() > 1) {
```

Comment:
reduce-scatter does not support non-uniform replica groups, and HLO verifier checks that, so I don't think this needs to be checked here.

Reply:
removed.